rust-runtime/aws-smithy-http-server/src/request/request_id.rs

```diff
@@ -123,25 +137,32 @@ impl<S> Layer<S> for ServerRequestIdProviderLayer {
     type Service = ServerRequestIdProvider<S>;
 
     fn layer(&self, inner: S) -> Self::Service {
-        ServerRequestIdProvider { inner }
+        ServerRequestIdProvider { inner, header_key: self.header_key.clone() }
     }
 }
 
 impl<Body, S> Service<http::Request<Body>> for ServerRequestIdProvider<S>
 where
-    S: Service<http::Request<Body>>,
+    S: Service<http::Request<Body>, Response = Response<crate::body::BoxBody>>,
+    S::Future: std::marker::Send + 'static,
 {
     type Response = S::Response;
     type Error = S::Error;
-    type Future = S::Future;
+    type Future = ServerRequestIdResponseFuture<S::Future>;
 
     fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
         self.inner.poll_ready(cx)
     }
 
     fn call(&mut self, mut req: http::Request<Body>) -> Self::Future {
-        req.extensions_mut().insert(ServerRequestId::new());
-        self.inner.call(req)
+        let request_id = ServerRequestId::new();
+        req.extensions_mut().insert(request_id.clone());
+        let header_key = self.header_key.clone();
+        ServerRequestIdResponseFuture {
+            request_id,
```

Comment:
In the case where `header_key` is `None` we don't need to insert the `request_id` here and hence we don't need to `request_id.clone()`.

Reply:
What if one wants to use the request ID in their handler?